File: RPG-Git/src/pokemon/Pokemon.java
package pokemon;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.Iterator;
import java.util.List;
import java.util.Map;
import java.util.Random;

/**
 * Pokemon class for all Pokemon types.
 * 
 * @author evinstk
 */
public class Pokemon {
    private static final int LEVEL_CAP = 100;
    private static final int EX_IV_MAX = 16;
    
    private final int index;
    private String name;
    private final Map<String, Short> base_stats;
    private final Map<String, Byte> iv;
    private Map<String, Integer> ev;
    int level;
    int exp;
    private List<Move> moves;
    final Type types[];
    final String experience_group;
    int hp;
    boolean fainted;
    Status status_ailment;
    
    /**
     * Constructs a Pokemon object according to name, level, and starting moves.
     * 
     * @param name name of the type of Pokemon to construct (e.g., "bulbasaur")
     * @param level level of the constructed Pokemon
     * @param move_names set of moves of the constructed Pokemon
     */
    public Pokemon(int index, int level, String[] move_names) {
        Random rand = new Random();
        HashMap<String, Byte> iv_builder = new HashMap<String, Byte>(4);
        String[] iv_keys = {"ATTACK", "DEFENSE", "SPEED", "SPECIAL"};
        String[] ev_keys = {"hp", "attack", "defense", "speed", "special"};

        this.index = index;
        
        name = pokemon.database.Query.getName(index);
        
        base_stats = pokemon.database.Query.getBaseStats(index);
        
        for (String key : iv_keys) {
            iv_builder.put(key, (byte)rand.nextInt(EX_IV_MAX));
            
        }
        byte iv_health = 0;
        //Simulates how health is determined in generation I games
        for (int i = 0; i < iv_builder.size(); i++) {
            byte[] adders = {8, 4, 2, 1};
            for (int j = 0; j < iv_keys.length; j++) {
                if (((String)iv_builder.keySet().toArray()[i] == iv_keys[j]) && 
                        (((byte)iv_builder.values().toArray()[i] % (byte)2) == 1)) {
                    iv_health += adders[j];
                }
            }
        }
        iv_builder.put("HP", iv_health);
        iv = Collections.unmodifiableMap(iv_builder);
        
        ev = new HashMap(5);
        for (String key : ev_keys) {
            ev.put(key, 0);
        }
        
        this.level = level;
        
        moves = new ArrayList<Move>(4);
        for (String move_name : move_names) {
            moves.add(new Move(1));
        }
        
        types = pokemon.database.Query.getType(index);
        
        experience_group = pokemon.database.Query.getExperienceGroup(name);
        
        initExp();
        
        hp = getStats().get("hp");
        
        fainted = false;
        
        status_ailment = Status.NORMAL;
    }
    
    public Map<String, Byte> getIV() {
        return iv;
    }
    
    public Map<String, Integer> getEV() {
        return ev;
    }
    
    public int getExp() {
        return exp;
    }
    
    /**
     * Initializes exp variable. Sets it to some value between the minimum
     * (inclusive) and maximum (exclusive) values for the current level.
     */
    private void initExp() {
        Random rand = new Random();
        exp = rand.nextInt(getMinExp(level + 1) - getMinExp(level)) + getMinExp(level);
    }
    
    /**
     * Takes a Pokemon level and returns minimum experience required to reach
     * it. The experience group is taken from the Pokemon's experience_group
     * variable.
     * 
     * @param level level to determine minimum experience for it
     * @return minimum experience required for the given level
     */
    private int getMinExp(int level){
        if (experience_group.equals("Fast")) {
            return ((4*((level)^3))/5);
        }
        else if (experience_group.equals("MediumFast")) {
            return ((level)^3);
        }
        else if (experience_group.equals("MediumSlow")) {
            return ((6/5)*((level)^3) - (15)*((level)^2) + (100)*(level) - 140);
        }
        else if (experience_group.equals("Slow")) {
            return (5*((level)^3)/4);
        }
        else {
            throw new Error("Unrecognized experience group");
        }
    }
    
    /**
     * Returns stats of the Pokemon. Stats are a function of base stats,
     * individual values, effort values, and level.
     * 
     * @return mapping of stat keys to values
     */
    public Map<String, Integer> getStats() {
        HashMap<String, Integer> ret = new HashMap<String, Integer>();
        
        Iterator iter = ev.keySet().iterator();
        while(iter.hasNext()) {
            String key = (String)iter.next();
            int value, m1, m2;
            if (key.equals("hp")) {
                m1 = 50;
                m2 = 10;
            }
            else {
                m1 = 0;
                m2 = 5;
            }
            value = (int) (((((iv.get(key.toUpperCase()) + base_stats.get(key.toUpperCase()) + (Math.sqrt(ev.get(key))/8) + m1) * level)/50) + m2));
            ret.put(key, value);
        }
        return ret;
    }
    
    /**
     * Returns current level of Pokemon. Needed outside of Pokemon object to
     * calculate damage.
     * 
     * @return Pokemon's level
     */
    public int getLevel() {
        return level;
    }
    
    /**
     * Returns type(s) of Pokemon. Needed for damage multiplier calculation.
     * 
     * @return Pokemon's types
     */
    public Type[] getTypes() {
        return types;
    }
    
    /**
     * Returns hp.
     * 
     * @return hp 
     */
    public int getHp() {
        return hp;
    }
    
    /**
     * Returns Pokemon's current moveset.
     * 
     * @return 
     */
    public List<Move> getMoves() {
        return moves;
    }
    
    /**
     * Reduces hp of Pokemon.
     * 
     * @param damage Damage dealt by attacking Pokemon 
     */
    public void takeDamage(int damage) {
        hp -= damage;
        if (hp <= 0) {
            hp = 0;
            fainted = true;
        }
    }
    
    public void gainExp(int exp) {
        this.exp = exp;
        checkLevelUp();
    }
    
    private void checkLevelUp() {
        if (exp >= getMinExp(level + 1)) {
            level++;
            checkLevelUp();
            return;
        }
        return;
    }
    
    public void setStatus(Status status_ailment) {
        this.status_ailment = status_ailment;
    }
    
    public Status getStatus() {
        return status_ailment;
    }
}

File: PokemonRedBlue/pokemon.cpp
/* 
 * File:   Pokemon.cpp
 * Author: evinstk
 *
 * Created on June 16, 2013, 8:19 PM
 */

#include "pokemon.h"
#include "query.h"
#include <cstdlib>
#include <iostream>
#include <time.h>
#include <math.h>

using namespace std;

Pokemon::Pokemon() {
    index = 0;
    name = new string("Empty");
    base_stats = NULL;
    iv = NULL;
    ev = NULL;
    level = 0;
    types = NULL;
    experience_group = -1;
    exp = 0;
    fainted = true;
    status_ailment = 0;
}

Pokemon::Pokemon(unsigned char index, int level) {
    this->index = index;
    
    name = query::getName(index);
    
    base_stats = query::getBaseStats(index);
    
    iv = new unsigned char[5];
    srand(time(NULL));
    for (size_t i = 1; i < sizeof(iv); i++) {
        iv[i] = (unsigned char)(rand() % 16);
    }
    int hp = 0;
    int adder = 8;
    for (size_t i = 0; i < 4; i++) {
        if ((iv[i + 1] % 2) == 1) {
            hp += adder;
        }
        adder /= 2;
    }
    iv[0] = hp;
    
    ev = new unsigned short[5];
    
    this->level = level;
    
    moves = query::getDefaultMoves(index);
    
    types = query::getTypes(index);
    
    experience_group = query::getExpGroup(index);
    
    initExp();
    
    this->hp = getHpStat();
    
    fainted = false;
    
    status_ailment = 0;
}

Pokemon::~Pokemon() {
    delete name;
    delete[] base_stats;
    delete[] iv;
    delete[] ev;
    delete[] types;
}

int Pokemon::getMinExp(unsigned char level) {
    if (experience_group == 0) {
        return ((4*((level)^3))/5);
    }
    else if (experience_group == 1) {
        return ((level)^3);
    }
    else if (experience_group == 2) {
        return ((6/5)*((level)^3) - (15)*((level)^2) + (100)*(level) - 140);
    }
    else if (experience_group == 3) {
        return (5*((level)^3)/4);
    }
}

#ifndef INIT_EXP
#define INIT_EXP
void Pokemon::initExp() {
    exp = rand() % (getMinExp(level + 1) - getMinExp(level)) + getMinExp(level);
}
#endif

string *Pokemon::getName() {
    return name;
}

unsigned char *Pokemon::getBaseStats() {
    return this->base_stats;
}

unsigned char *Pokemon::getIv() {
    return this->iv;
}

unsigned short *Pokemon::getEv() {
    return this->ev;
}

unsigned char Pokemon::getHp() {
    return hp;
}

unsigned char Pokemon::getStat(unsigned char stat_index) {
    unsigned char base_stat = getBaseStats()[stat_index];
    unsigned char individual_value = getIv()[stat_index];
    unsigned short effort_value = getEv()[stat_index];
    unsigned char curr_level = level;
    float m0, m1;
    if (stat_index == 0) {
        m0 = 50.f;
        m1 = 10.f;
    }
    else {
        m0 = 0.f;
        m1 = 5.f;
    }
    
    unsigned char ret = (unsigned char)((((float)(individual_value) + (float)(base_stat) + ((float)(sqrt(effort_value))/8.f) + m0) * (float)(curr_level)) / 50.f + m1);
    return ret;
}

unsigned char Pokemon::getHpStat() {
    return getStat(0);
}

unsigned char Pokemon::getAttackStat() {
    return getStat(1);
}

unsigned char Pokemon::getDefenseStat() {
    return getStat(2);
}

unsigned char Pokemon::getSpeedStat() {
    return getStat(3);
}

unsigned char Pokemon::getSpecialStat() {
    return getStat(4);
}

bool Pokemon::isFainted() {
    return fainted;
}
